File: android/src/main/kotlin/dev/moaz/dash_bubble/src/BubbleOptions.kt
package dev.moaz.dash_bubble.src

import android.os.Parcelable
import io.flutter.plugin.common.MethodCall
import kotlinx.parcelize.Parcelize

/** BubbleOptions is the class that contains the options for the bubble.
 * @param bubbleIcon The icon of the bubble.
 * @param closeIcon The icon of the close button.
 * @param startLocationX The x coordinate of the bubble.
 * @param startLocationY The y coordinate of the bubble.
 * @param bubbleSize The size of the bubble.
 * @param opacity The opacity of the bubble.
 * @param enableClose Whether the bubble can be closed or not.
 * @param closeBehavior The behavior of the close button.
 * @param distanceToClose The distance between the bubble and the bottom edge of the screen to show the close button.
 * @param enableAnimateToEdge Whether the bubble can be animated to the edge or not.
 * @param enableBottomShadow Whether the bubble has a bottom shadow behind the close button or not.
 * @param keepAliveWhenAppExit Whether the bubble will be kept alive when the app is closed or not.
 */
@Parcelize
data class BubbleOptions(
    val bubbleIcon: String?,
    val closeIcon: String?,
    val startLocationX: Double?,
    val startLocationY: Double?,
    val bubbleSize: Double?,
    val opacity: Double?,
    val enableClose: Boolean?,
    val closeBehavior: Int?,
    val distanceToClose: Double?,
    val enableAnimateToEdge: Boolean?,
    val enableBottomShadow: Boolean?,
    val keepAliveWhenAppExit: Boolean?,
    val socketUrl: String?,
    val orderManagementUrl: String?,
    val userToken: String?,
    val userId: Int?
) : Parcelable {
    companion object {
        /**
         * Get the bubble options from the method call.
         * @param call The method call.
         * @return The bubble options.
         */
        fun fromMethodCall(call: MethodCall): BubbleOptions {
            return BubbleOptions(
                bubbleIcon = call.argument(Constants.BUBBLE_ICON),
                closeIcon = call.argument(Constants.CLOSE_ICON),
                startLocationX = call.argument(Constants.START_LOCATION_X),
                startLocationY = call.argument(Constants.START_LOCATION_Y),
                bubbleSize = call.argument(Constants.BUBBLE_SIZE),
                opacity = call.argument(Constants.OPACITY),
                enableClose = call.argument(Constants.ENABLE_CLOSE),
                closeBehavior = call.argument(Constants.CLOSE_BEHAVIOR),
                distanceToClose = call.argument(Constants.DISTANCE_TO_CLOSE),
                enableAnimateToEdge = call.argument(Constants.ENABLE_ANIMATE_TO_EDGE),
                enableBottomShadow = call.argument(Constants.ENABLE_BOTTOM_SHADOW),
                keepAliveWhenAppExit = call.argument(Constants.KEEP_ALIVE_WHEN_APP_EXIT),
                socketUrl = call.argument(Constants.SOCKET_URL),
                orderManagementUrl = call.argument(Constants.ORDER_URL),
                userToken = call.argument(Constants.USER_TOKEN),
                userId = call.argument(Constants.USER_ID),
            )
        }
    }
}

File: lib/src/models/bubble_options.dart
import '../enums/enums.dart';

/// [BubbleOptions] is a class that contains all the customizable options for the bubble passed to the [DashBubble.startBubble] method.
class BubbleOptions {
  /// The icon of the bubble.
  final String? bubbleIcon;

  /// The icon of the close button.
  final String? closeIcon;

  /// The initial starting position of the bubble on the x axis.
  final double startLocationX;

  /// The initial starting position of the bubble on the y axis.
  final double startLocationY;

  /// The size of the bubble.
  final double bubbleSize;

  /// The opacity of the bubble and the close button.
  final double opacity;

  /// Whether to show the close button or not.
  final bool enableClose;

  final CloseBehavior closeBehavior;

  /// The distance between the bubble and the bottom edge of the screen to show the close button.
  final double distanceToClose;

  /// Whether to animate the bubble to the edge of the screen when it is dragged to the edge of the screen or not.
  final bool enableAnimateToEdge;

  /// Whether to show the bottom shadow behind the close button of the bubble or not.
  final bool enableBottomShadow;

  /// Whether to keep the bubble alive when the app is closed or not.
  final bool keepAliveWhenAppExit;

  /// socket url if we want to connect and listen to socket events in background
  final String? socketUrl;

  /// if we are sending socket then we need user token
  final String? userToken;

  final String? orderManagementBaseURL;

  final int? userId;

  BubbleOptions(
      {this.bubbleIcon,
      this.closeIcon,
      this.startLocationX = 0,
      this.startLocationY = 200,
      this.bubbleSize = 60,
      this.opacity = 1,
      this.enableClose = true,
      this.closeBehavior = CloseBehavior.following,
      this.distanceToClose = 100,
      this.enableAnimateToEdge = true,
      this.enableBottomShadow = true,
      this.keepAliveWhenAppExit = false,
      this.socketUrl,
      this.userToken,
      this.orderManagementBaseURL,
      this.userId})
      : assert(
          bubbleSize >= 0,
          'bubbleSize must be greater than or equal to 0',
        ),
        assert(
          opacity >= 0 && opacity <= 1,
          'opacity must be between 0 and 1',
        );

  Map<String, dynamic> toMap() {
    return {
      'bubbleIcon': bubbleIcon,
      'closeIcon': closeIcon,
      'startLocationX': startLocationX,
      'startLocationY': startLocationY,
      'bubbleSize': bubbleSize,
      'opacity': opacity,
      'enableClose': enableClose,
      'closeBehavior': closeBehavior.index,
      'distanceToClose': distanceToClose,
      'enableAnimateToEdge': enableAnimateToEdge,
      'enableBottomShadow': enableBottomShadow,
      'keepAliveWhenAppExit': keepAliveWhenAppExit,
      'socketUrl': socketUrl,
      'userToken': userToken,
      'userId': userId,
      "orderUrl": orderManagementBaseURL
    };
  }
}
